File: solutions/hash_table/1-5.py
"""
Problem Number 1.5
Problem : Given two strings, write an algo to check if they are one edit (insert, remove, replace) away.
"""

import array

class check_string (object):

	def check_insert(self, str1, str2):
		index1=0
		index2=0

		while (index1 < len(str1)) and (index2 < len(str2)):
			if str1[index1] != str2[index2]:
				index2+=1
				if index2 != index1 + 1:
					print ("Insert False")
					return
				continue
			index2+=1
			index1+=1

		print ("Insert True")			

	def check_replace(self, str1, str2):
		is_changed = False;
		for i in range (len(str1)):
			if str1[i] != str2[i]:
				if is_changed == True:
					print ("Replace False")
					return
				is_changed = True
		print ("Replace True")

	def check_remove(self, str1, str2):
		index1 = 0
		index2 = 0 
		while (index1 < len(str1)) and (index2 < len(str2)):
			if str1[index1] != str2[index2]:
				index1+=1
				if index1 > (index2+1) :
					print ("Remove False")
					return
				continue
			index1+=1
			index2+=1

		print("Remove True")

print ("Please enter the 1st string : ")
s1 = input()

print ("Please enter the 2nd string : ")
s2 = input()

check_string_obj = check_string()

# Its replaced if the lengths are same
if len(s1) == len(s2):
	check_string_obj.check_replace(s1,s2)

elif len(s1) == len(s2)+1:
	check_string_obj.check_remove(s1,s2)

elif len(s1)+1 == len(s2) :
	check_string_obj.check_insert(s1,s2)

else :
	print ("False")
	

File: solutions/hash_table/1-5.cpp
/*
Problem Number 1.5
Problem : Given two strings, write an algo to check if they are one edit (insert, remove, replace) away.
*/

#include <iostream>
#include <streambuf>
using namespace std;

class MyString
{
	public:
				void checkInsert(string s1, string s2)
				{
					int index1 = 0, index2 = 0;

					while (index1 < s1.size() && index2 < s2.size())
					{

						if ((index2 - index1) > 1)
						{
							cout << "\nInsert False";
							return;
						}

						if (s1[index1] != s2[index2])
						{
							index2++;
							continue;
						}


						index1++;
						index2++;
					}

					cout << "\nInsert True";
				}

				void checkReplace(string s1, string s2)
				{
					bool is_replaced = false;

					for (int i = 0; i < s1.size(); i++)
					{
						if (s1[i] != s2[i])
						{
							if (is_replaced)
							{
								cout << "\nReplaced False";
								return;
							}
							is_replaced = true;
						}
					}
					cout << "\nReplaced True";
				}

				void checkRemove(string s1, string s2)
				{
					int index1 = 0, index2 = 0;

					while (index1 < s1.size() && index2 < s2.size())
					{
						if ((index1 - index2) > 1)
						{
							cout << "\nRemove False";
							return;
						}

						if (s1[index1] != s2[index2])
						{
							index1++;
							continue;
						}
						index1++;
						index2++;

					}

					cout << "Remove True";
				}
};


int main()
{
	string s1, s2;
	cout << "\nPlease enter enter the first string : ";
	cin >> s1;

	cout << "Please enter the second string : ";
	cin >> s2;

	MyString MS;

	// If replaced
	if (s1.size() == s2.size())
		MS.checkReplace(s1, s2);
	// If inserted
	else if ((s1.size() + 1) == s2.size())
		MS.checkInsert(s1, s2);
	// If removed
	else if (s1.size() == (s2.size() + 1))
		MS.checkRemove(s1, s2);
	else
		cout << "\nFalse";
}
